### ISSUE TITLE
Memory leak in dyn_str_new

### ISSUE THREAD

**Issue by gy741** [opencog/link-grammar#651]
Hello.

I found a Memory leak in link-grammar.

This problem is detected by valgrind.

Please confirm.

Thanks.

Summary: Memory leak 
OS: CentOS 7 64bit
Version: commit commit 9684460
PoC Download:  [leak_dyn_str_new.zip](https://github.com/opencog/link-grammar/files/1653938/leak_dyn_str_new.zip)

Steps to reproduce:
1.Download the .POC files.
2.Execute the following command
    : valgrind --leak-check=full  ./link-parser < $PoC

```
==681== 274 (24 direct, 250 indirect) bytes in 1 blocks are definitely lost in loss record 2 of 2
==681==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==681==    by 0x148BCA: dyn_str_new (utilities.c:435)
==681==    by 0x1189CB: display_word_split (print-dict.c:222)
==681==    by 0x1189CB: dict_display_word_info (print-dict.c:389)
==681==    by 0x10F71B: x_issue_special_command (command-line.c:319)
==681==    by 0x10F71B: issue_special_command (command-line.c:546)
==681==    by 0x10D9DA: special_command (link-parser.c:460)
==681==    by 0x10D9DA: main (link-parser.c:751)
==681== 
==681== LEAK SUMMARY:
==681==    definitely lost: 24 bytes in 1 blocks
==681==    indirectly lost: 250 bytes in 1 blocks
==681==      possibly lost: 0 bytes in 0 blocks
==681==    still reachable: 0 bytes in 0 blocks
==681==         suppressed: 0 bytes in 0 blocks
==681== 
==681== For counts of detected and suppressed errors, rerun with: -v
==681== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

```
=================================================================
==29413==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 24 byte(s) in 1 object(s) allocated from:
    #0 0x4dd170 in __interceptor_malloc (/home/karas/link-grammar/run/bin/link-parser+0x4dd170)
    #1 0x618c42 in dyn_str_new /home/karas/link-grammar/link-grammar/utilities.c:435:16
    #2 0x51ec71 in x_issue_special_command /home/karas/link-grammar/link-parser/command-line.c:319:9
    #3 0x51ec71 in issue_special_command /home/karas/link-grammar/link-parser/command-line.c:546
    #4 0x516976 in special_command /home/karas/link-grammar/link-parser/link-parser.c:460:3
    #5 0x516976 in main /home/karas/link-grammar/link-parser/link-parser.c:751
    #6 0x7f574cab41c0 in __libc_start_main /build/glibc-itYbWN/glibc-2.26/csu/../csu/libc-start.c:308

Indirect leak of 250 byte(s) in 1 object(s) allocated from:
    #0 0x4dd170 in __interceptor_malloc (/home/karas/link-grammar/run/bin/link-parser+0x4dd170)
    #1 0x618c7c in dyn_str_new /home/karas/link-grammar/link-grammar/utilities.c:438:12
    #2 0x51ec71 in x_issue_special_command /home/karas/link-grammar/link-parser/command-line.c:319:9
    #3 0x51ec71 in issue_special_command /home/karas/link-grammar/link-parser/command-line.c:546
    #4 0x516976 in special_command /home/karas/link-grammar/link-parser/link-parser.c:460:3
    #5 0x516976 in main /home/karas/link-grammar/link-parser/link-parser.c:751
    #6 0x7f574cab41c0 in __libc_start_main /build/glibc-itYbWN/glibc-2.26/csu/../csu/libc-start.c:308

SUMMARY: AddressSanitizer: 274 byte(s) leaked in 2 allocation(s).
```

==========
[Acknowledgement]
This work was supported by ICT R&D program of MSIP/IITP. [R7518-16-1001, Innovation hub for high Performance Computing]

**Comment by linas** [comment]
Yes, I see this.

**Comment by linas** [comment]
I just pushed a fix for this in 639cd0f0171f364a355c68b2eb7a5cd68b1d7a1c

Thanks!  this is interesting.
